zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/AbstractInterpreter.java

```diff
@@ -17,71 +17,330 @@
 
 package org.apache.zeppelin.interpreter;
 
-import org.apache.commons.text.StringSubstitutor;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.apache.zeppelin.interpreter.xref.*;
+import org.apache.zeppelin.interpreter.xref.annotation.Experimental;
+import org.apache.zeppelin.interpreter.xref.annotation.ZeppelinApi;
 import org.apache.zeppelin.interpreter.thrift.InterpreterCompletion;
-import org.apache.zeppelin.resource.ResourcePool;
-import org.apache.zeppelin.resource.ResourcePoolMap;
+import org.apache.zeppelin.scheduler.SchedulerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * Interface for interpreters.
+ * If you want to implement new Zeppelin interpreter, extend this class
+ *
+ * Please see,
+ * https://zeppelin.apache.org/docs/latest/development/writingzeppelininterpreter.html
+ *
+ * open(), close(), interpret() is three the most important method you need to implement.
+ * cancel(), getProgress(), completion() is good to have
+ * getFormType(), getScheduler() determine Zeppelin's behavior
+ */
+public abstract class AbstractInterpreter implements Interpreter {
```

Comment:
Seems safe to assume that this is not original code but moved from existing implementation?